app/src/main/java/com/android/sample/model/workout/WorkoutRepositoryFirestore.kt

```diff
@@ -0,0 +1,139 @@
+package com.android.sample.model.workout
+
+import android.util.Log
+import com.google.firebase.Firebase
+import com.google.firebase.auth.auth
+import com.google.firebase.firestore.DocumentSnapshot
+import com.google.firebase.firestore.FirebaseFirestore
+import com.squareup.moshi.JsonDataException
+import com.squareup.moshi.JsonEncodingException
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
+import dev.zacsweers.moshix.sealed.reflect.MoshiSealedJsonAdapterFactory
+import kotlin.reflect.full.companionObject
+import kotlin.reflect.full.companionObjectInstance
+
+class WorkoutRepositoryFirestore<T : Workout>(
+    private val db: FirebaseFirestore,
+    private val clazz: Class<T>
+) : WorkoutRepository<T> {
+
+  fun getDocumentName(): String {
+    val companionObject = clazz.kotlin.companionObjectInstance
+    val documentNameProperty =
+        clazz.kotlin.companionObject?.members?.find { it.name == "DOCUMENT_NAME" }
+    return documentNameProperty?.call(companionObject) as? String
+        ?: throw IllegalStateException("DOCUMENT_NAME not found for class ${clazz.simpleName}")
+  }
```

Comment:
This is rather ugly, but I guess we can't do better since there's no static members in Kotlin